### PR TITLE
Validate metadata payload schemas

### PR DESCRIPTION
**Pull request by brenetic**
## Validate metadata request object schemas

We want to validate the metadata for a given service against the schemas
to make sure that the payload is correct. If not then we raise a 422.

## Add schemas necessary for text field

https://trello.com/c/i01rmxHP/1088-implement-text-field-component-in-the-metadata-api

Co-authored-by: Natalie Seeto <natalie.seeto@digital.justice.gov.uk>

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -23,6 +23,7 @@ class ApplicationController < ActionController::API
   end
 
   before_action AuthenticateApplication
+  before_action :validate_schema
```
Comment (tomas-stefano) [inline app/controllers/application_controller.rb:26]: One way to do it wihout doing the class self on the filter and adding a method the controller:

```
before_action ValidateSchema
```

The good thing is that Rails pass the controller to you.
Then on ValidateSchema you can move the logic there:

```ruby
class ValidateSchema
  def self.before(controller)
    return unless controller.request.post?
     schema = "request.#{controller.request.params['controller']}"
     validate(controller.request.params, schema)
  rescue JSON::Schema::ValidationError, JSON::Schema::SchemaError, SchemaNotFoundError => e
     controller.render '...'
  end

  # ...
end
```

What do you think?

File: app/filters/validate_schema.rb [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    end
+
+    def validate(metadata, schema)
+      service_schema = find(schema)
+      JSON::Validator.validate!(service_schema, metadata)
+    end
+
+    def find(schema)
+      schema_file = schema.gsub('.', '/')
+      schema = File.read(File.join(
```
Comment (tomas-stefano) [inline app/filters/validate_schema.rb:18]: Since the schemas won't change per service, is it a good idea to cache the File read into a constant or a class variable? *If yes, that's okay to do it in another PR I believe.

File: schemas/definition/widthclass.json [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+      ],
+      "category": [
+        "htmlattributes"
+      ]
+    }
+  },
+  "category": [
+    "definition"
+  ]
+}
```
Comment (tomas-stefano) [inline schemas/definition/widthclass.json:34]: We talked about this today, sorry if I don't remember. Is it okay as camel case this file, for now, right?

**Review by tomas-stefano — APPROVED** [review]
Excellent work and effort! I left minor comments but doesn't alter the PR. I am happy to talk tomorrow over a call if necessary. All good 🚀 🎸

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (brenetic) [reply app/controllers/application_controller.rb:26]: yup good idea! will update

File: schemas/definition/widthclass.json (hunk 3 above)
Comment (brenetic) [reply schemas/definition/widthclass.json:34]: do you mean `widthclass` should be `width_class`?

File: app/filters/validate_schema.rb (hunk 2 above)
Comment (brenetic) [reply app/filters/validate_schema.rb:18]: so at this point the schema it is looking for is a request schema. there are currently two types, one for a service and one for a version. the request schemas reference the service schemas internally. we could cache the request.service schema and the request.version schema though . . .

File: schemas/definition/widthclass.json (hunk 3 above)
Comment (tomas-stefano) [reply schemas/definition/widthclass.json:34]: Yes, `widthClass` and `enumMap`

**Review by tomas-stefano — APPROVED** [review]
Awesome work 🚀 🎸
